### PR TITLE
contract drive rest services with vertx

### PR DESCRIPTION
**Pull request by pmlopes**
This is a new article for the blog.

This time i cover the subject of contract driven API, design, implementation and testing. As with previous blog entries it is a light reading just showing how to get things done.

This should answer questions like we have in the forum about integrate tools such as:

* Swagger
* RAML
* Mashery
* Apiary
* etc...

The article covers RAML but i think with some research readers who prefer something else can achive the same...

### CODE REVIEW

**Comment by pmlopes** [comment]
@cescoffier, @purplefox, @vietj please review.

File: src/site/blog/posts/vertx3-contract-driven-rest-services.md [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+see services being offered as REST APIs that we are allowed to consume and we already see this trend from the big names
+in the industry, e.g.: [Facebook](https://developers.facebook.com/), [Wikipedia](https://www.mediawiki.org/wiki/API:Main_page),
+[Amazon](https://developer.amazon.com/public/apis), [Twitter](https://dev.twitter.com/overview/documentation),
+[Google](https://developers.google.com/apis-explorer/#p/) and [Reddit](https://www.reddit.com/dev/api) they all offer
+APIs.
+
+Of course making an REST API using Vert.x is quite simple, just use [Vert.x Web](http://vertx.io/docs/#web) and you can
+start writing your API in seconds, however an API without documentation is not an API since no developer will know how
+to use it. However this is one of the most sensitive issues to tackle in software development, nobody likes to write
+documentation. However, in the REST age where REST-based webservices are obiquitous, documentation for public
```
Comment (cescoffier) [inline src/site/blog/posts/vertx3-contract-driven-rest-services.md]: ubiquitous ?

File: src/site/blog/posts/vertx3-contract-driven-rest-services.md [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+webservices is a necessity. There are a lot of tools out there, two of the most popular are without a doubt:
+
+ * [Swagger](http://swagger.io)
+ * [RAML](http://raml.org)
+
+Both frameworks have a large ecosystem of tools and tooling around but they tackle the documentation from two different
+perspectives. While Swagger is purely a documentation tool it tackles the problem from bottom up and RAML does document
+top down. They both rely on a document (JSON for Swagger), (YAML for RAML).
```
Comment (cescoffier) [inline src/site/blog/posts/vertx3-contract-driven-rest-services.md]: (JSON for Swagger, YAML for RAML)

File: src/site/blog/posts/vertx3-contract-driven-rest-services.md [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+/hello:
+  get:
+    responses:
+      200:
+        body:
+          application/json:
+```
+
+So if you didn't understood why I named RAML as a top down documentation tool, I think it becomes clear now. So there
+are some basic definition on the top of the file like, `title`, `baseUri` and `versio` which should be sef explanatory.
```
Comment (cescoffier) [inline src/site/blog/posts/vertx3-contract-driven-rest-services.md]: `version`

Comment (cescoffier) [inline src/site/blog/posts/vertx3-contract-driven-rest-services.md]: self

File: src/site/blog/posts/vertx3-contract-driven-rest-services.md [hunk 4]
```diff
@@ -0,0 +1,220 @@
...
+
+And then we start with the API documentation, so we specify that at the `URL` `/hello` using the HTTP verb `GET` you are
+expected to get a response with status code `200` and the body of the response should have content type
+`application/json`. This is a very minimal document, one could go over and specify the `json schema` for the response,
```
Comment (cescoffier) [inline src/site/blog/posts/vertx3-contract-driven-rest-services.md]: can you put a link to the json schema specification

File: src/site/blog/posts/vertx3-contract-driven-rest-services.md [hunk 5]
```diff
@@ -0,0 +1,220 @@
...
+  }
+
+  @Test
+  public void testHelloEndpoint() {
+  }
+}
+```
+
+So at this moment you have a simple test, I'll share the code of the runner class (basically it just instantiates a
+`Vertx` instance and load the verticle we just implemented above) and has a empty test `testHelloEndpoint`.
```
Comment (cescoffier) [inline src/site/blog/posts/vertx3-contract-driven-rest-services.md]: You can link to the blog post explaining how to run tests.

File: src/site/blog/posts/vertx3-contract-driven-rest-services.md [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+
+Tests run: 1, Failures: 0, Errors: 0, Skipped: 0
+```
+
+And we are done, we now have a API that follow the contract, you can now keep developing your API and implementation and
+have a test driven approach to be sure that the contract is not broken.
+
+## Article source code
+
+You can get the full source code for this article [here](https://github.com/pmlopes/vert-x3-experiments/tree/experiments/raml).
```
Comment (cescoffier) [inline src/site/blog/posts/vertx3-contract-driven-rest-services.md]: just a raw question, did you try their "console"

**Comment by pmlopes** [comment]
I'll integrate the web console with the example and add some text about it

**Comment by michel-kraemer** [comment]
So, the PR is now ready to be merged?

**Comment by cescoffier** [comment]
Yes.

**Comment by michel-kraemer** [comment]
@ vertx core developers: should I update the vert-x3.github.io repository (do I even have the permissions) or will you do it?

**Comment by pmlopes** [comment]
A new site build is required, the draft flag was still active. I've removed it form master

**Comment by cescoffier** [comment]
I will do it.

**Comment by cescoffier** [comment]
Done, sorry for the delay (finding a good enough network in airports is not that simple)

**Pull request merged**